File: Golang/Medium/3021_Alice_and_Bob_Playing_Flower_Game/solution.go
package aliceandbobplayingflowergame

type state struct {
	n int
	m int
}

func flowerGame(n int, m int) int64 {
	return _flowerGame(n, m, make(map[state]int64))
}

func _flowerGame(n int, m int, memo map[state]int64) int64 {
	if n == 0 || m == 0 {
		return int64(0)
	}

	key := state{n, m}
	if v, exist := memo[key]; exist {
		return v
	}

	count := int64(0)
	if (n+m)%2 != 0 {
		count += 1
	}

	count += _flowerGame(n-1, m, memo)
	count += _flowerGame(n, m-1, memo)
	count -= _flowerGame(n-1, m-1, memo)
	memo[key] = count
	return count
}

File: Python/Medium/3021_Alice_and_Bob_Playing_Flower_Game/solution.py
class Solution:
    def flowerGame(self, n: int, m: int) -> int:
        return Solution._flowerGame(n, m, {})
      
    def _flowerGame(n, m, memo):
        if n == 0 or m == 0:
            return 0

        key = (n, m)
        if key in memo:
            return memo[key]
        
        count = 0
        if (n + m) % 2 != 0:
            count += 1

        count += Solution._flowerGame(n-1, m, memo)
        count += Solution._flowerGame(n, m-1, memo)
        count -= Solution._flowerGame(n-1, m-1, memo)

        memo[key] = count
        return count
